runners/flink/src/test/java/org/apache/beam/runners/flink/translation/wrappers/streaming/io/UnboundedSourceWrapperTest.java

```diff
@@ -626,13 +650,30 @@ public void testSourceWithNoReaderDoesNotShutdown() throws Exception {
 
       try {
         thread.start();
-        List<UnboundedSource.UnboundedReader<KV<Integer, Integer>>> localReaders =
-            sourceWrapper.getLocalReaders();
-        while (localReaders != null && !localReaders.isEmpty()) {
-          Thread.sleep(200);
-          // should stay alive
-          assertThat(thread.isAlive(), is(true));
+        // Wait to see if the wrapper shuts down immediately in case it doesn't have readers
+        if (!shouldHaveReaders) {
+          // The expected state is for finalizeSource to sleep instead of exiting
+          while (true) {
+            if (thread.getStackTrace().length >= 2) {
```

Comment:
Interestingly the test passes with this change, though it isn't clear why :)

Please switch the lines and do the length check as `callStack.length` (can also be combined into existing condition).

Reply:
Just did that.